### PR TITLE
[ray_client] actors v0

### LINKED ISSUE #11976: Support actors

### PR DESCRIPTION
**Pull request by barakmich**
## Related issue number

Depends on #12141 
Closes #11976

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ericl** [comment]
Needs rebase / pull of master.

**Comment by barakmich** [comment]
Of course it does. I mean, I did say "Depends on #12141" -- which has now been merged

File: python/ray/experimental/client/client_app.py [hunk 1]
```diff
@@ -3,6 +3,27 @@
...
+
+    def say_hello(self, whom):
+        self.count += 1
+        return ("Hello " + whom, self.count)
+
+
+actor = HelloActor.remote()
+s, count = ray.get(actor.say_hello.remote("you"))
+print(s, count)
+assert s == "hello you"
```
Comment (ericl) [inline python/ray/experimental/client/client_app.py:19]:
```suggestion
assert s == "Hello you"
```

File: python/ray/experimental/client/client_app.py [hunk 2]
```diff
@@ -3,6 +3,27 @@
...
+assert count == 1
+s, count = ray.get(actor.say_hello.remote("world"))
+print(s, count)
+assert s == "hello world"
```
Comment (ericl) [inline python/ray/experimental/client/client_app.py:23]:
```suggestion
assert s == "Hello world"
```

File: python/ray/experimental/client/client_app.py [hunk 3]
```diff
@@ -3,6 +3,27 @@
...
+assert count == 2
```
Comment (ericl) [inline python/ray/experimental/client/client_app.py:24]: One thing I noticed is that passing actor handles crashes trying to serialize the handle:

```
@ray.remote
def f(handle):
    return handle.say_hello.remote("Bob")

ray.get(f.remote(actor))
```

File: python/ray/experimental/client/common.py [hunk 4]
```diff
@@ -27,12 +38,64 @@ def __call__(self, *args, **kwargs):
...
+
+
+class ClientActorHandle:
+    def __init__(self, actor_id: ClientActorRef,
+                 actor_class: ClientActorClass):
+        self.actor_id = actor_id
+        self.actor_class = actor_class
+
+    def __getattr__(self, key):
+        return ClientRemoteMethod(self, key)
```
Comment (ericl) [inline python/ray/experimental/client/common.py:77]: It seems __getattr__ is intercepting too much here, e.g., if you try to pickle this handle it ends up calling __getattr__ which returns a remote method.

**Review by ericl — COMMENTED** [review]
Just tried it out, works well. Found a corner case with the getattr handling ^^

File: python/ray/experimental/client/client_app.py (hunk 1 above)
Comment (AmeerHajAli) [reply python/ray/experimental/client/client_app.py:19]: Is this passing?

File: python/ray/experimental/client/client_app.py [hunk 5]
```diff
@@ -3,6 +3,27 @@
 ray.connect("localhost:50051")
 
 
+@ray.remote
+class HelloActor:
+    def __init__(self):
+        self.count = 0
+
+    def say_hello(self, whom):
```
Comment (AmeerHajAli) [inline python/ray/experimental/client/client_app.py:11]: type annotation please?

File: python/ray/experimental/client/server/server.py [hunk 6]
```diff
@@ -68,24 +71,62 @@ def WaitObject(self, request, context=None) -> ray_client_pb2.WaitResponse:
...
+        return ray_client_pb2.ClientTaskTicket(return_id=actor_ref.binary())
+
+    def _schedule_function(self, task, context=None):
         if task.payload_id not in self.function_refs:
             funcref = self.object_refs[task.payload_id]
             func = ray.get(funcref)
             if not isinstance(func, ClientRemoteFunc):
-                raise Exception("Attempting to schedule something that "
+                raise Exception("Attempting to schedule function that "
                                 "isn't a ClientRemoteFunc")
```
Comment (AmeerHajAli) [inline python/ray/experimental/client/server/server.py:122]:
```suggestion
                                "isn't a ClientRemoteFunc.")
```

File: python/ray/experimental/client/server/server.py [hunk 7]
```diff
@@ -68,24 +71,62 @@ def WaitObject(self, request, context=None) -> ray_client_pb2.WaitResponse:
...
+        return ray_client_pb2.ClientTaskTicket(return_id=output.binary())
+
+    def _schedule_actor(self, task, context=None):
+        with stash_api_for_tests(self._test_mode):
+            if task.payload_id not in self.registered_actor_classes:
+                actor_class_ref = self.object_refs[task.payload_id]
+                actor_class = ray.get(actor_class_ref)
+                if not inspect.isclass(actor_class):
+                    raise Exception("Attempting to schedule actor that "
+                                    "isn't a ClientActorClass")
```
Comment (AmeerHajAli) [inline python/ray/experimental/client/server/server.py:106]:
```suggestion
                                    "isn't a ClientActorClass.")
```

File: python/ray/experimental/client/server/server.py [hunk 8]
```diff
@@ -68,24 +71,62 @@ def WaitObject(self, request, context=None) -> ray_client_pb2.WaitResponse:
...
+        if actor_handle is None:
+            raise Exception(
+                "Can't run an actor the server doesn't have a handle for")
+        arglist = _convert_args(task.args)
+        with stash_api_for_tests(self._test_mode):
+            output = getattr(actor_handle, task.name).remote(*arglist)
+            self.object_refs[output.binary()] = output
+        return ray_client_pb2.ClientTaskTicket(return_id=output.binary())
+
+    def _schedule_actor(self, task, context=None):
```
Comment (AmeerHajAli) [inline python/ray/experimental/client/server/server.py:99]: type annotation?

File: python/ray/experimental/client/worker.py [hunk 9]
```diff
@@ -96,21 +102,62 @@ def wait(self,
...
+        if (inspect.isfunction(function_or_class)
+                or is_cython(function_or_class)):
+            return ClientRemoteFunc(function_or_class)
+        elif inspect.isclass(function_or_class):
+            return ClientActorClass(function_or_class)
+        else:
+            raise TypeError("The @ray.remote decorator must be applied to "
+                            "either a function or to a class.")
+
+    def call_remote(self, instance, kind, *args, **kwargs):
```
Comment (AmeerHajAli) [inline python/ray/experimental/client/worker.py:117]: type annotation?

**Review by AmeerHajAli — COMMENTED** [review]
LGTM.
I am able to reproduce the bug Eric mentioned on passing an actor handler to a function.
I also tested a bunch of other combinations of arguments (e.g., passing functions/reference as arguments) but these seem to work. What also does not work is passing an actor handler to an actor (to any function in the actor).
Can we have another test in addition to the basic one that tests a bunch of combinations of passing references/functions/actors?

**Comment by barakmich** [comment]
Addressed and rebased -- passing an ActorHandle around is kind of bleh. I mean, I'm forming a plan think I can do it (and, similarly, functions) in a reasonable way by tomorrow but that's some 'fun' behavior.

**Comment by ericl** [comment]
@barakmich is there something here beyond fixing the getattr problem? I believe the way normal Ray actor handles work is by building an actor method table. We can probably do something similar to override just those methods instead of getattr, which is preventing serialization of those handles.

**Comment by ericl** [comment]
@AmeerHajAli what do you think about merging this as part1 of actors--- fixing actor handles can be part2

**Comment by ericl** [comment]
I think we can merge this as part 1 of actors... lint is failing though:
```

++++command -v clang-format
++++/home/travis/build/ray-project/tmp.sv31RqRIQV/ci/travis/check-git-clang-format-output.sh
Running clang-format against branch 234df9091e57f2a4d7574e2ae53635bb5c862b9e, with hash 234df9091e57f2a4d7574e2ae53635bb5c862b9e
clang-format failed:
diff --git a/src/ray/protobuf/ray_client.proto b/src/ray/protobuf/ray_client.proto
index 7f70460c5..fd8fe5345 100644
--- a/src/ray/protobuf/ray_client.proto
+++ b/src/ray/protobuf/ray_client.proto
@@ -29,7 +29,6 @@ message Arg {
   Type type = 4;
 }```

**Comment by AmeerHajAli** [comment]
We can split it to two PRs, can we open an issue for this though so we can make sure to track it properly?

**Comment by barakmich** [comment]
SGTM,  found the lint that only `--all` caught

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Looks good, let's merge and iterate on the fixes

**Pull request merged**